crates/core/app/tests/common/mod.rs

```diff
@@ -83,19 +87,130 @@ impl TempStorageExt for TempStorage {
+                    trace!(%name, %enabled, %sequence_number, "injecting validator into app state");
+                }
+                // ...or print a warning if there are not any validators.
+                if v.is_empty() {
+                    warn!("`with_penumbra_auto_app_state` was called but builder has no validators")
+                }
+            });
+
+        // Add the validators to the app state.
+        let app_state: AppState = match app_state {
+            AppState::Checkpoint(_) => anyhow::bail!("checkpoint app state isn't supported"),
+            AppState::Content(mut content) => {
+                // Inject the builder's validators into the staking component's genesis state.
+                std::mem::replace(
+                        &mut content.stake_content.validators,
+                        validators
+                    )
+                    .tap(|overwritten| {
+                        // Log a warning if this overwrote any validators already in the app state.
+                        if !overwritten.is_empty() {
+                            warn!(?overwritten, "`with_penumbra_auto_app_state` overwrote validators in the given AppState")
+                        }
+                    });
+                AppState::Content(content)
+            }
+        };
+
+        // Serialize the app state into bytes, and add it to the builder.
+        serde_json::to_vec(&app_state)
+            .map_err(Self::Error::from)
+            .map(|s| self.app_state(s))
+    }
+
+    fn with_penumbra_single_validator(self) -> Self {
+        use {
+            decaf377_rdsa::VerificationKey,
+            penumbra_keys::keys::{SpendKey, SpendKeyBytes},
+            penumbra_proto::{
+                core::keys::v1::{GovernanceKey, IdentityKey},
+                penumbra::core::component::stake::v1::Validator,
+            },
+            rand::Rng,
+            rand_core::OsRng,
+        };
+
+        // Generate a spend authoration key.
+        let bytes = {
+            let spend_key = SpendKey::from(SpendKeyBytes(OsRng.gen()));
+            let spend_auth_key = spend_key.spend_auth_key();
+            let verification_key = VerificationKey::from(spend_auth_key);
+            verification_key.to_bytes()
+        };
+
+        // Generate a validator entry using the generated key.
+        let validator = Validator {
+            identity_key: Some(IdentityKey {
+                ik: bytes.to_vec().clone(),
+            }),
+            governance_key: Some(GovernanceKey {
+                // NB: for now, we will use the same key for governance. See the documentation of
+                // `GovernanceKey` for more information about cold storage of validator keys.
+                gk: bytes.to_vec().clone(),
+            }),
+            consensus_key: {
+                let signing_key = ed25519_consensus::SigningKey::new(OsRng);
+                signing_key.as_bytes().as_slice().to_vec()
```

Comment:
This is publishing the validator's signing (private) key, rather than their verification (public) key. Presumably this isn't causing problems at the moment because we're not actually producing valid headers, but we should rethink this API to better handle key management.

Reply:
xref https://github.com/penumbra-zone/penumbra/pull/3902#issuecomment-1971674791

Reply:
marking this as resolved via https://github.com/penumbra-zone/penumbra/pull/3902#issuecomment-1972239643 and the referenced conversation in #testing.

we found a better way forward.

(_and also importantly, the verification key is used. thanks for calling that out!_)